libc/src/math/generic/exp2m1f.cpp

```diff
@@ -0,0 +1,144 @@
+    }
+
+    // x <= -25
+    if (x <= -25.0f) {
+      // 2^(-inf) - 1 = -1
+      if (xbits.is_inf())
+        return -1.0f;
+      // 2^nan - 1 = nan
+      if (xbits.is_nan())
+        return x;
+
+      int rounding = fputil::quick_get_round();
+      if (rounding == FE_UPWARD || rounding == FE_TOWARDZERO)
+        return -0x1.ffff'fep-1f; // -1.0f + 0x1.0p-24f
+
+      fputil::set_errno_if_required(ERANGE);
+      fputil::raise_except_if_required(FE_UNDERFLOW);
+      return -1.0f;
+    }
+  }
+
+  // For -25 < x < 128, to compute 2^x, we perform the following range
+  // reduction: find hi, mid, lo such that:
+  //   x = hi + mid + lo, in which:
+  //     hi is an integer,
+  //     0 <= mid * 2^5 < 32 is an integer,
+  //     -2^(-6) <= lo <= 2^(-6).
+  // In particular,
+  //   hi + mid = round(x * 2^5) * 2^(-5).
+  // Then,
+  //   2^x = 2^(hi + mid + lo) = 2^hi * 2^mid * 2^lo.
+  // 2^mid is stored in the lookup table of 32 elements.
+  // 2^lo is computed using a degree-4 minimax polynomial generated by Sollya.
+  // We perform 2^hi * 2^mid by simply add hi to the exponent field of 2^mid.
+
+  // kf = (hi + mid) * 2^5 = round(x * 2^5)
+  float kf;
+  int k;
+#ifdef LIBC_TARGET_CPU_HAS_NEAREST_INT
+  kf = fputil::nearest_integer(x * 32.0f);
+  k = static_cast<int>(kf);
+#else
+  constexpr float HALF[2] = {0.5f, -0.5f};
+  k = static_cast<int>(fputil::multiply_add(x, 32.0f, HALF[x < 0.0f]));
+  kf = static_cast<float>(k);
+#endif // LIBC_TARGET_CPU_HAS_NEAREST_INT
+
+  // dx = lo = x - (hi + mid) = x - kf * 2^(-5)
+  double dx = fputil::multiply_add(-0x1.0p-5f, kf, x);
+
+  // hi = floor(kf * 2^(-4))
+  // exp_hi = shift hi to the exponent field of double precision.
+  int64_t exp_hi =
+      static_cast<int64_t>(static_cast<uint64_t>(k >> ExpBase::MID_BITS)
+                           << fputil::FPBits<double>::FRACTION_LEN);
+  // mh = 2^hi * 2^mid
+  // mh_bits = bit field of mh
+  int64_t mh_bits = ExpBase::EXP_2_MID[k & ExpBase::MID_MASK] + exp_hi;
+  double mh = fputil::FPBits<double>(static_cast<uint64_t>(mh_bits)).get_val();
+
+  // Degree-4 polynomial approximating (2^x - 1)/x generated by Sollya with:
+  // > display = hexadecimal;
+  // > fpminimax((2^x - 1)/x, 4, [|D...|], [-2^-6, 2^-6]);
+  constexpr double COEFFS[5] = {0x1.62e42fefa39efp-1, 0x1.ebfbdff8131c4p-3,
+                                0x1.c6b08d7061695p-5, 0x1.3b2b1bee74b2ap-7,
+                                0x1.5d88091198529p-10};
+  double dx_sq = dx * dx;
+  double c1 = fputil::multiply_add(dx, COEFFS[0], 1.0);
+  double c2 = fputil::multiply_add(dx, COEFFS[2], COEFFS[1]);
+  double c3 = fputil::multiply_add(dx, COEFFS[4], COEFFS[3]);
+  double p = fputil::multiply_add(dx_sq, c3, c2);
+  // 2^x = 2^(hi + mid + lo)
+  //     = 2^(hi + mid) * 2^lo
+  //     ~ mh * (1 + lo * P(lo))
+  //     = mh + (mh*lo) * P(lo)
+  double exp2_lo = fputil::multiply_add(p, dx_sq, c1);
+  return static_cast<float>(fputil::multiply_add(exp2_lo, mh, -1.0));
```

Comment:
I don't get how we end up with `mh * (1 + lo * P(lo))`. The code in `exp2f_impl.h` was:

```cpp
  double dx_sq = dx * dx;
  double c1 = fputil::multiply_add(dx, COEFFS[0], 1.0);
  double c2 = fputil::multiply_add(dx, COEFFS[2], COEFFS[1]);
  double c3 = fputil::multiply_add(dx, COEFFS[4], COEFFS[3]);
  double p = fputil::multiply_add(dx_sq, c3, c2);
  // 2^x = 2^(hi + mid + lo)
  //     = 2^(hi + mid) * 2^lo
  //     ~ mh * (1 + lo * P(lo))
  //     = mh + (mh*lo) * P(lo)
  return static_cast<float>(fputil::multiply_add(p, dx_sq * mh, c1 * mh));
```

My understanding is that we should have:

```
c1 = a0*lo + 1
c2 = a2*lo + a1
c3 = a4*lo + a3
P(lo) = lo^2 * c3 + c2
      = lo^2 * (a4*lo + a3) + (a2*lo + a1)
      = a4*lo^3 + a3*lo^2 + a2*lo + a1
2^lo ~ a4*lo^4 + a3*lo^3 + a2*lo^2 + a1*lo + a0
     = (a4*lo^3 + a3*lo^2 + a2*lo + a1) * lo + a0
     = a0 + lo*P(lo)
2^x = 2^(hi + mid + lo)
    = 2^(hi + mid) * 2^lo
    ~ mh * (a0 + lo*P(lo))
    = (mh*a0) + (mh*lo) * P(lo)
    = (mh*a0) + (mh*lo) * (a4*lo^3 + a3*lo^2 + a2*lo + a1)
    = mh*a4*lo^4 + mh*a3*lo^3 + mh*a2*lo^2 + mh*a1*lo + mh*a0
```

whereas you have:

```
2^x ~ (mh*(lo^2)) * P(lo) + (mh*c1)
    = (mh*(lo^2)) * (a4*lo^3 + a3*lo^2 + a2*lo + a1) + (mh * (a0*lo + 1))
    = mh*a4*lo^5 + mh*a3*lo^4 + mh*a2*lo^3 + mh*a1*lo^2 + mh*a0*lo + mh
```

Reply:
The confusion is that `COEFFS` array was generated for `(2^x - 1)/x`, so `a4*lo^4 + a3*lo^3 + a2*lo^2 + a1*lo + a0` is `~ (2^lo - 1)/lo`, not `2^lo.

Reply:
Oops. :(

Reply:
That's why I was stepping back and forth a bit on stating the degree of the polynomial.  Technically it is a degree-4 from the Sollya command, but it is actually a degree-5 polynomial approximating `2^lo`.

Reply:
Ah, it makes sense now.

Reply:
Oh, so example 3 in section 4.1.1 of *Elementary Functions* was actually related. It all makes sense now.